syncer/syncer.go

```diff
@@ -0,0 +1,402 @@
+		fetcher:       fetcher,
+		syncState:     notSynced,
+		syncTimer:     time.NewTicker(conf.SyncInterval),
+		awaitSyncedCh: make(chan struct{}),
+		shutdownCtx:   shutdownCtx,
+		cancelFunc:    cancel,
+	}
+}
+
+// Close stops the syncing process and the goroutines syncer spawns.
+func (s *Syncer) Close() {
+	// TODO: ensure goroutines are all terminated before shutting down
+	s.cancelFunc()
+}
+
+// AwaitSynced returns the channel that will be notified whenever the node enters synced state.
+func (s *Syncer) AwaitSynced() chan struct{} {
+	s.awaitSyncedMu.Lock()
+	defer s.awaitSyncedMu.Unlock()
+	return s.awaitSyncedCh
+}
+
+// ListenToGossip returns true if the node is listening to gossip for blocks/TXs/ATXs data.
+func (s *Syncer) ListenToGossip() bool {
+	return s.conf.AlwaysListen || s.getSyncState() >= gossipSync
+}
+
+// IsSynced returns true if the nodes is fully in sync with the network.
+func (s *Syncer) IsSynced(ctx context.Context) bool {
+	res := s.getSyncState() == synced
+	s.logger.WithContext(ctx).With().Info("node sync state",
+		log.Bool("synced", res),
+		log.FieldNamed("current", s.ticker.GetCurrentLayer()),
+		log.FieldNamed("latest", s.mesh.LatestLayer()),
+		log.FieldNamed("validated", s.mesh.ProcessedLayer()))
+	return res
+}
+
+// Start starts the main sync loop that tries to sync data for every SyncInterval.
+func (s *Syncer) Start(ctx context.Context) {
+	s.syncOnce.Do(func() {
+		if s.ticker.GetCurrentLayer() == 0 {
+			s.setSyncState(ctx, synced)
+		}
+		for {
+			select {
+			case <-s.shutdownCtx.Done():
+				s.logger.WithContext(ctx).Info("stopping sync to shutdown")
+				return
+			case <-s.syncTimer.C:
+				s.synchronize(ctx)
+			}
+		}
+	})
+}
+
+// ForceSync manually start a sync process outside the main sync loop. If the node is already running a sync process,
+// ForceSync will be ignored.
+func (s *Syncer) ForceSync(ctx context.Context) {
+	s.logger.WithContext(ctx).Debug("executing ForceSync")
+	go s.synchronize(ctx)
+}
+
+func (s *Syncer) isClosed() bool {
+	select {
+	case <-s.shutdownCtx.Done():
+		return true
+	default:
+		return false
+	}
+}
+
+func (s *Syncer) getSyncState() syncState {
+	return (syncState)(atomic.LoadUint32((*uint32)(&s.syncState)))
+}
+
+func (s *Syncer) setSyncState(ctx context.Context, newState syncState) {
+	oldState := syncState(atomic.SwapUint32((*uint32)(&s.syncState), uint32(newState)))
+	if oldState != newState {
+		s.logger.WithContext(ctx).With().Info("sync state change",
+			log.String("from", oldState.String()),
+			log.String("to", newState.String()),
+			log.FieldNamed("current", s.ticker.GetCurrentLayer()),
+			log.FieldNamed("latest", s.mesh.LatestLayer()),
+			log.FieldNamed("validated", s.mesh.ProcessedLayer()))
+		events.ReportNodeStatusUpdate()
+		// notify subscribes
+		s.awaitSyncedMu.Lock()
+		defer s.awaitSyncedMu.Unlock()
+		if newState == synced {
+			close(s.awaitSyncedCh)
+		} else if oldState == synced {
+			s.awaitSyncedCh = make(chan struct{})
```

Comment:
since you are replacing the channel maybe we should change how `awaitSynced` works because this switch might cause other processes to listen to the wrong channel

Reply:
you are right.
changed to use client supplied channel for notification.
activation.go wants to select on a channel so i don't want to make AwaitSynced() block.